### DISCUSSION TITLE
Sleigh: how to write bit manipulation instructions

### DISCUSSION THREAD

**Discussion by esaulenka** [NationalSecurityAgency/ghidra#5966] [Q&A]
I am working on improving Renesas M16C support. It have a lot of bit manipulation opcodes, like BTST, BSET, BCLR, BNOT, BAND, BOR, etc.
All of these have similar syntax, allowing to deal with several operands: bit N in regM, bit N at address M, bit N at address (M + reg) etc (16 possible variants in total).

So, the common approach is to write specific table for all possible operands and reuse it several times in opcodes.
But looks like syntax `{ local tmp = R0[imm8:1,1]; export tmp; }` (where is `imm8` is just a constant) is not allowed.
I can rewrite it as `{ local tmp:1 = (R0 >> imm8) & 1; export tmp;	 }` but such syntax allows only read bit, not write it back.

Could anyone advice how to implement it without writing tons of similar code?

**Comment by GhidorahRex** [comment]
You won't be able to read/write from a sub-constructor. You'll have to export the operand and width in separate constructors. then create a read/write mask in the main constructor. Similar to what the 68000 does for it's bit manipulation instructions.

**Comment by esaulenka** [comment]
Nice idea, thanks.

In short, if anyone else will be interested:

- declare byte offset:
https://github.com/esaulenka/ghidra_m16c/blob/2153c218b5847a21d278ad464fc55cd7a482f709/data/languages/m16c.slaspec#L268-L285
```
s_bitV: bitNo8, bitBase8[SB] is SB & op8_11=10; bitNo8 & bitBase8     { local addr = SB + bitBase8; export *[near]:1 addr; }
s_bitV: bitNo8, bitBase8[FB] is FB & op8_11=11; bitNo8 & bitBase8     { local addr = FB + bitBase8; export *[near]:1 addr; }
s_bitV: bitNo16, bitBase16     is op8_11=15; bitNo16 & bitBase16      { export *[near]:1 bitBase16; }
```

- declare bit mask:
https://github.com/esaulenka/ghidra_m16c/blob/2153c218b5847a21d278ad464fc55cd7a482f709/data/languages/m16c.slaspec#L287-L299
```
s_bitB: bitNo8, bitBase8[SB] is SB & op8_11=10; bitNo8 & bitBase8     { local mask:1 = 1<<bitNo8; export mask; }
s_bitB: bitNo8, bitBase8[FB] is FB & op8_11=11; bitNo8 & bitBase8     { local mask:1 = 1<<bitNo8; export mask; }
s_bitB: bitNo16, bitBase16     is op8_11=15; bitNo16 & bitBase16      { local mask:1 = 1<<bitNo16; export mask; }
```

-  and use it:
```
:BSET s_bitV                     is (op0_7=0x7E & op12_15=0x9)... & s_bitB & s_bitV {
	s_bitV = s_bitV | s_bitB;
}
:BTST s_bitV                     is (op0_7=0x7E & op12_15=0xB)... & s_bitV & s_bitB {
	local tmp = (s_bitV & s_bitB);
	Cflg = (tmp != 0);
	Zflg = (tmp == 0);
}
```

@GhidorahRex, maybe you also can advice how to improve it. In some cases this code doesn't create references in the Disassembler window, but in decompiled code it looks well:
```
7e 9f 6a 15     BSET       0x2, 0x2ad
7e 9f a0 8a     BSET       0x0, BYTE_near_1154
```
It uses these slaspec pieces:
```
define token data16 (16)
	bitNo16 = (0,2)
	bitBase16 = (3,15)
;
s_bitV: bitNo16, bitBase16     is op8_11=15; bitNo16 & bitBase16      { export *[near]:1 bitBase16; }
s_bitB: bitNo16, bitBase16     is op8_11=15; bitNo16 & bitBase16      { local mask:1 = 1<<bitNo16; export mask; }

:BSET s_bitV                     is (op0_7=0x7E & op12_15=0x9)... & s_bitB & s_bitV {
	s_bitV = s_bitV | s_bitB;
}
```
